### PR TITLE
add case for mem pagelock and shared pages cases

### PR DESCRIPTION
**Pull request by nanli1**
VIRT-297169: Memory pagelocking and shared pages
Signed-off-by: nanli <nanli@redhat.com>

Depend on 
- https://github.com/avocado-framework/avocado-vt/pull/3776

### CODE REVIEW

**Comment by nanli1** [comment]
```
avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 memory.backing.page_locking_and_shared_pages
 (01/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.page_default.ksm_default.lock_without_hard_limit: PASS (55.84 s)
 (02/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.page_default.ksm_default.lock_hard_limit: PASS (54.87 s)
 (03/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.page_default.ksm_disabled.lock_default: PASS (56.71 s)
 (04/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.page_default.ksm_disabled.lock_without_hard_limit: PASS (60.20 s)
 (05/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.page_default.ksm_disabled.lock_hard_limit: PASS (61.16 s)
 (06/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.hugepage.ksm_default.lock_without_hard_limit: PASS (60.68 s)
 (07/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.hugepage.ksm_default.lock_hard_limit: PASS (56.88 s)
 (08/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.hugepage.ksm_disabled.lock_default: PASS (60.80 s)
 (09/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.hugepage.ksm_disabled.lock_without_hard_limit: PASS (55.48 s)
 (10/10) type_specific.io-github-autotest-libvirt.memory.backing.page_locking_and_shared_pages.memory_allocation.hugepage.ksm_disabled.lock_hard_limit: PASS (57.81 s)

```

**Comment by nanli1** [comment]
modular_redesign

**Review by cliping — APPROVED** [review]
LGTM

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    def get_init_vm_attrs():
+        """
+        Get vm attrs.
+
+        :return expected vm attrs dict.
+        """
+        mb_value = ""
+        for item in [lock_dict, ksm_dict, hugepages_dict]:
+            if item != "":
+                mb_value = mb_value+item+","
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:37]: mb_value += "%s," % item

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            test.fail("Fail to consume guest memory. Error:%s" % output)
+        session.close()
+
+        test.log.info("TEST_STEP6:  Check the locked memory pages")
+        lock_pages_2 = int(utils_memory.read_from_vmstat("nr_mlock"))
+        nr_unevictable_2 = int(utils_memory.read_from_vmstat("nr_unevictable"))
+
+        if lock_config == "lock_default":
+            if lock_pages != lock_pages_2:
+                test.fail("Nr_mlock should not change to:%d", lock_pages_2)
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:96]: , -> '%'

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            if nr_unevictable != nr_unevictable_2:
+                test.fail("Unevictable should not change to:%d", nr_unevictable_2)
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:98]: , -> '%'

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            return
+        else:
+            if page_config == "page_default":
+                if (nr_unevictable_2 - nr_unevictable)*4 < mem_value:
+                    test.fail("Unevictable memory pages should less than %d", mem_value)
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:103]: , -> '%'

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                if (lock_pages_2 - lock_pages)*4 < mem_value:
+                    test.fail("Locked memory pages should less than %d", mem_value)
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:105]: , -> '%'

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            elif page_config == "hugepage":
+                if (nr_unevictable_2 - nr_unevictable)*4 < 0:
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:107]: Do we need `*4 `?

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                    test.fail("Check the unevictable memory failed")
+                if (lock_pages_2 - lock_pages)*4 < 0:
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:109]: Do we need *4 ?

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            return
+        else:
+            if page_config == "page_default":
+                if (nr_unevictable_2 - nr_unevictable)*4 < mem_value:
+                    test.fail("Unevictable memory pages should less than %d", mem_value)
+                if (lock_pages_2 - lock_pages)*4 < mem_value:
+                    test.fail("Locked memory pages should less than %d", mem_value)
+            elif page_config == "hugepage":
+                if (nr_unevictable_2 - nr_unevictable)*4 < 0:
+                    test.fail("Check the unevictable memory failed")
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:108]: Could you give more descriptive error message? Like expect the memory should be xxx, but ...

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+                if (lock_pages_2 - lock_pages)*4 < 0:
+                    test.fail("Check the lock pages failed")
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:110]: Same as above

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+        test.log.info("TEST_STEP7: Check the qemu process limit of memory lock")
+        cmd = "prlimit -p `pidof qemu-kvm` -l |awk '/MEMLOCK/ {print $6,$7}'"
+        stdout = process.run(cmd, shell=True).stdout_text.strip()
+        if stdout != process_limit+" "+process_limit:
+            test.fail("Not get expected:%s" % (process_limit+process_limit))
```
Comment (dzhengfy) [inline libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:116]: Expect to be ` process_limit+" "+process_limit` ?
Is it better to reuse below?
`expected =  "%s %s" % (process_limit, process_limit)`

Comment (dzhengfy) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:116]: It is better to reuse libvirt_memory.get_qemu_process_memlock_hard_limit() after enhancement

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 1 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 2 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 3 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 4 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 6 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  It's just along with the polarion case designed , Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 7 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 8 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 9 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks dzheng,  Updated

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 10 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:119]: Thanks dzheng,  This `libvirt_memory.get_qemu_process_memlock_hard_limit()` returns the hard value ,we need both soft and hard value here ,So I just use` process.run()`

Comment (dzhengfy) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py:119]: Then could you add a new function like below to return both soft and hard limits? For get_qemu_process_memlock_hard_limit(), it should be replaced with your new function. Other references need to be updated. Thanks.

`def get_qemu_process_memlock_limit():
    """
    :return: tuple, (soft_limit, hard_limit)
    """
     cmd = "prlimit -p `pidof qemu-kvm` -l |awk '/MEMLOCK/ {print $6,$7}'"
     stdout = process.run(cmd, shell=True).stdout_text.strip().split()
    return (stdout[0], stdout[1])
`

**Review by dzhengfy — COMMENTED** [review]
See above

File: libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py (hunk 10 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/page_locking_and_shared_pages.py]: Thanks for your advice , Here is the new pr, https://github.com/avocado-framework/avocado-vt/pull/3776 , Other references need to be updated in another pr later

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**